### PR TITLE
ISPN-4390 JSR-166 ConcurrentHashMap Realignment

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-4390
https://issues.jboss.org/browse/ISPN-6933

Cross-port the ConcurrentHashMap from Doug Lea's JSR-166 repo (specifically from the src/jdk8/java/util/concurrent path, revision 1.3)
Also include some necessary classes: Helper, ThreadLocalRandom
Extract the Unsafe logic to an independent class
Also rename the package from jdk8backports to jsr166mods to avoid confusion

### CODE REVIEW

**Comment by slaskawi** [comment]
This test failure might be related: http://ci.infinispan.org/viewLog.html?buildId=41593&buildTypeId=bt9&guest=1#testNameId143828433828339114

```
java.lang.NullPointerException
    at org.infinispan.functional.FunctionalDistributionTest.lambda$iterate$515(FunctionalDistributionTest.java:101)
    at org.infinispan.functional.FunctionalDistributionTest$$Lambda$634/865376728.apply(Unknown Source)
    at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
    at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
    at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
    at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374)
    at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:512)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:502)
    at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
    at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
    at org.infinispan.functional.FunctionalDistributionTest.iterate(FunctionalDistributionTest.java:102)
    at org.infinispan.functional.FunctionalDistributionTest.doTestDistribution(FunctionalDistributionTest.java:80)
    at org.infinispan.functional.FunctionalDistributionTest.testDistributionFromPrimaryOwner(FunctionalDistributionTest.java:46)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:497)
    at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:84)
    at org.testng.internal.Invoker.invokeMethod(Invoker.java:714)
    at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:901)
    at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1231)
    at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:127)
    at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:111)
    at org.testng.TestRunner.privateRun(TestRunner.java:767)
    at org.testng.TestRunner.run(TestRunner.java:617)
    at org.testng.SuiteRunner.runTest(SuiteRunner.java:348)
    at org.testng.SuiteRunner.access$000(SuiteRunner.java:38)
    at org.testng.SuiteRunner$SuiteWorker.run(SuiteRunner.java:382)
    at org.testng.internal.thread.ThreadUtil$2.call(ThreadUtil.java:64)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
------- Stdout: -------
[TestSuiteProgress] Test starting: org.infinispan.distribution.groups.TransactionalGetGroupKeysTest.testGetKeysInGroupWithPersistenceAndSkipCacheLoader
[TestSuiteProgress] Test failed: org.infinispan.functional.FunctionalDistributionTest.testDistributionFromPrimaryOwner
20:41:11,002 ERROR (testng-FunctionalDistributionTest) [TestSuiteProgress] Test failed: org.infinispan.functional.FunctionalDistributionTest.testDistributionFromPrimaryOwner
java.lang.NullPointerException
    at org.infinispan.functional.FunctionalDistributionTest.lambda$iterate$515(FunctionalDistributionTest.java:101)
    at org.infinispan.functional.FunctionalDistributionTest$$Lambda$634/865376728.apply(Unknown Source)
    at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
    at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
    at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
    at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374)
    at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:512)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:502)
    at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)[TestSuiteProgress] Tests succeeded: 458, failed: 1, skipped: 0
[TestSuiteProgress] Test starting: org.infinispan.functional.distribution.rehash.FunctionalNonTxJoinerBecomingBackupOwnerTest.testBackupOwnerJoiningDuringPutIfAbsent
```

**Comment by tristantarrant** [comment]
Unlikely because it doesn't fail locally nor on the 8.2.x cherry-pick

**Comment by slaskawi** [comment]
Let me retrigger CI

**Comment by galderz** [comment]
CI still shows that failure

**Comment by slaskawi** [comment]
The new build didn't start (there are some problems with TC Master)

On Fri, Aug 12, 2016 at 11:54 AM, Galder Zamarreño <notifications@github.com

**Comment by galderz** [comment]
@tristantarrant Needs rebasing

File: commons/src/main/java/org/infinispan/commons/util/concurrent/jsr166mods/ThreadLocalRandom.java [hunk 1]
```diff
@@ -0,0 +1,1044 @@
...
+ * secure.  Consider instead using {@link java.security.SecureRandom}
+ * in security-sensitive applications. Additionally,
+ * default-constructed instances do not use a cryptographically random
+ * seed unless the {@linkplain System#getProperty system property}
+ * {@code java.util.secureRandomSeed} is set to {@code true}.
+ *
+ * @since 1.7
+ * @author Doug Lea
+ */
+public class ThreadLocalRandom extends Random {
```
Comment (wburns) [inline commons/src/main/java/org/infinispan/commons/util/concurrent/jsr166mods/ThreadLocalRandom.java]: Can we not use java.util.concurrent.ThreadLocalRandom?  Or it needs to reference package level methods probably?  If we add this we will have the class loader leak issue still.

**Comment by tristantarrant** [comment]
I messed up the rebase, so the changes have been lost. I'll fix it by tomorrow

**Comment by tristantarrant** [comment]
Will reopen a new one

**Comment by Sanne** [comment]
Was a new PR ever sent?

I re-opened https://issues.jboss.org/browse/ISPN-6933

**Comment by danberindei** [comment]
@Sanne The `jdk8backported` package was removed. We now use Caffeine for the bounded data container and regular `java.util.concurrent.ConcurrentHashMap` for the unbounded one.

**Comment by Sanne** [comment]
yes, thanks I realised later. Sorry for the question, the reason was that the above issue was in the limbo status of "PR sent" .. all good now.

**Pull request closed without merging**